google/cloud/storage/examples/storage_bucket_acl_samples.cc

```diff
@@ -0,0 +1,111 @@
+// Copyright 2018 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "google/cloud/storage/client.h"
+#include <functional>
+#include <iostream>
+#include <map>
+#include <sstream>
+
+namespace {
+struct Usage {
+  std::string msg;
+};
+
+char const* ConsumeArg(int& argc, char* argv[]) {
+  if (argc < 2) {
+    return nullptr;
+  }
+  char const* result = argv[1];
+  std::copy(argv + 2, argv + argc, argv + 1);
+  argc--;
+  return result;
+}
+
+std::string command_usage;
+
+void PrintUsage(int argc, char* argv[], std::string const& msg) {
+  std::string const cmd = argv[0];
+  auto last_slash = std::string(cmd).find_last_of('/');
+  auto program = cmd.substr(last_slash + 1);
+  std::cerr << msg << "\nUsage: " << program << " <command> [arguments]\n\n"
+            << "Commands:\n"
+            << command_usage << std::endl;
+}
+
+void ListBucketAcl(google::cloud::storage::Client client, int& argc,
+                   char* argv[]) {
+  if (argc != 2) {
+    throw Usage{"list-bucket-acl <bucket-name>"};
+  }
+  auto bucket_name = ConsumeArg(argc, argv);
+  //! [list bucket acl] [START storage_print_bucket_acl]
```

Comment:
What is this?

Reply:
The `//! [tag]` comments are parsed by Doxygen to produce code snippets for each function, for example:

https://googlecloudplatform.github.io/google-cloud-cpp/0.2.0/storage/classgoogle_1_1cloud_1_1storage_1_1v0_1_1Client.html#ad2b12e1262905f3c66e910bb944b237c

or just the [class documentation](https://googlecloudplatform.github.io/google-cloud-cpp/0.2.0/storage/classgoogle_1_1cloud_1_1storage_1_1v0_1_1Client.html) because the bookmark ids change.

The `[START tag]` comments are used to extract the google cloud documentation, for example this:

https://cloud.google.com/storage/docs/listing-buckets#storage-list-buckets-go


The C++ code snippets are on in the Google documentation (yet).